kvserver: add per-operation lock reliability settings

stevendanna opened the pull request:
Preserving unreplicated locks during split, merge, and lease transfers have different trade offs. For instance, during a split all lock updates are done in memory without any new replicated writes, whereas for merge and lease transfers requiring replicating locks through raft.

Here, we put the different operations under different settings since we may want to ship different defaults for the different operations.

Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/143127)
<!-- Reviewable:end -->

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143127#-:-OLjGwnaArPUCmJGmzlr:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @stevendanna)



<!-- Sent from Reviewable.io -->

stevendanna committed:
kvserver: add per-operation lock reliability settings

Preserving unreplicated locks during split, merge, and lease transfers have
different trade offs. For instance, during a split all lock updates
are done in memory without any new replicated writes, whereas for
merge and lease transfers requiring replicating locks through raft.

Here, we put the different operations under different settings since
we may want to ship different defaults for the different operations.

Epic: none
Release note: None

stevendanna commented:
Test failure potentially addressed here: https://github.com/cockroachdb/cockroach/pull/143357

stevendanna commented:
The most recent failure is perhaps known and unrelated to this PR. I'm going to bors this.

stevendanna commented:
bors r=yuzefovich
